File: Java/575. Distribute Candies.java
/**
 * https://leetcode.com/problems/distribute-candies/
 */

import java.util.HashSet;
import java.util.Set;

class Solution {
    // O(n) time complexity
    // O(n) space complexity
    public int distributeCandies(int[] candyType) {
        int len = candyType.length;
        int max = len / 2;
        Set<Integer> set = new HashSet<>();
        for (int ct : candyType) {
            set.add(ct);
        }
        return Math.min(max, set.size());
    }
}

File: Kotlin/575. Distribute Candies.kt
/**
 * https://leetcode.com/problems/distribute-candies/
 */

class Solution {
    // O(n) time complexity
    // O(n) space complexity
    fun distributeCandies(candyType: IntArray): Int {
        val set = mutableSetOf<Int>()
        candyType.forEach {
            set.add(it)
        }
        return Math.min(set.size, candyType.size / 2)
    }
}